### PR TITLE
Investments: benchmarks

### LINKED ISSUE #918: Benchmarks for `pallet-investments`
Once #909 is merged, we need to implement benchmarks for the `pallet-investments`

### PR DESCRIPTION
**Pull request by lemunozm**
# Description

Fixes #918

## Changes and Descriptions

- [x] Benchmark implementation
- [x] Passing with `MockRuntime`
- [x] Passing with `altair` and `centrifuge` runtimes.
  - [x] **Current issues getting the correct `investment_id` from a pool. I think [this](https://github.com/centrifuge/centrifuge-chain/pull/1558/files#diff-80b949673789ef15cf42a6c1b1e63a0ff4287ce47d312fb574add40f4f553cdeR37) from #1558 could solve the issue.**
- [x] Updated weights in runtimes
- [x] Extrinsic using the correct weights from benchmarks.

## Benchmarking results (already normalized to our node times)

- `update_invest_order()` &
- `update_invest_redeem()`:
  - `94_000_000 + reads(8) + writes(4)`
- `collect_investments()` &
- `collect_investments_for()` &
- `collect_redemptions()` &
- `collect_redemptions_for()`:
  - `97_000_000 + 4_000_000 * N + reads(7) + reads(1) * N + writes(3)`

### CODE REVIEW

File: pallets/investments/src/benchmarking.rs [hunk 1]
```diff
@@ -10,3 +10,134 @@
...
+	T::InvestmentId: Default + Into<CurrencyOf<T>>,
+	T::Accountant: PoolBenchmarkHelper<AccountId = T::AccountId>
+		+ InvestmentIdBenchmarkHelper<
+			InvestmentId = T::InvestmentId,
+			PoolId = <T::Accountant as PoolBenchmarkHelper>::PoolId,
+		>,
+	<T::Accountant as PoolBenchmarkHelper>::PoolId: Default + Copy,
+{
+	fn get_investment_id() -> T::InvestmentId
+where {
```
Comment (wischli) [inline pallets/investments/src/benchmarking.rs:40]: Nit: The `where` can be removed
```suggestion
	fn get_investment_id() -> T::InvestmentId {
```

File: pallets/investments/src/benchmarking.rs [hunk 2]
```diff
@@ -10,3 +10,134 @@
...
+			+ InvestmentIdBenchmarkHelper<
+				InvestmentId = T::InvestmentId,
+				PoolId = <T::Accountant as PoolBenchmarkHelper>::PoolId,
+			>,
+		<T::Accountant as PoolBenchmarkHelper>::PoolId: Default + Copy,
+)]
+mod benchmarks {
+	use super::*;
+
+	#[benchmark]
```
Comment (wischli) [inline pallets/investments/src/benchmarking.rs:64]: Just wanted to express that writing v2 benchmarks is definitely more enjoyable to me. Awesome that we can make use of `?`. Hope you feel the same 🙃

File: runtime/altair/src/lib.rs [hunk 3]
```diff
@@ -1745,7 +1745,7 @@ impl<
 			),
 		};
 
-		if is_tranche_investor {
+		if is_tranche_investor || cfg!(feature = "runtime-benchmarks") {
```
Comment (wischli) [inline runtime/altair/src/lib.rs:1748]: Very smart! 🎉

**Review by wischli — APPROVED** [review]
Awesome work @lemunozm! Super clean and easy to follow, loving it.

File: pallets/investments/src/benchmarking.rs (hunk 2 above)
Comment (lemunozm) [reply pallets/investments/src/benchmarking.rs:64]: Yes! When I realized I could remove the `.unwrap()` everywhere it was awesome!

My only concern is what I wrote in the `eng-rust` channel.

File: libs/types/src/tokens.rs [hunk 4]
```diff
@@ -183,7 +183,18 @@ where
 /// enables us to use the `TrancheCurrency` type separately where solely this
 /// enum variant would be relevant. Most notably, in the `struct Tranche`.
 #[derive(
-	Clone, Copy, PartialOrd, Ord, PartialEq, Eq, Debug, Encode, Decode, TypeInfo, MaxEncodedLen,
+	Clone,
+	Copy,
+	Default,
```
Comment (mustermeiszer) [inline libs/types/src/tokens.rs:188]: I do not like default feature on this. Can we hide it behind `std` or `runtime-benchmarks`?

File: libs/test-utils/src/mocks/accountant.rs [hunk 5]
```diff
@@ -166,6 +166,27 @@ macro_rules! impl_mock_accountant {
...
+			#[cfg(feature = "runtime-benchmarks")]
+			impl<Tokens> cfg_traits::benchmarking::PoolBenchmarkHelper for $name<Tokens> {
+				type AccountId = $account_id;
+				type Balance = $balance;
+				type PoolId = ();
+
+				fn bench_create_pool(_: Self::PoolId, _: &Self::AccountId) {}
+
+				fn bench_investor_setup(_: Self::PoolId, _: Self::AccountId, _: Self::Balance) {}
+			}
```
Comment (mustermeiszer) [inline libs/test-utils/src/mocks/accountant.rs:178]: Why do we have this on the accountant and not on the `pool`?

Comment (lemunozm) [reply libs/test-utils/src/mocks/accountant.rs:178]: The actual benchmark implementation is already in the pool. But we need here too, to pass the benchmarking tests using the accountant mock (which actually mocks the pool from the accountant's perspective).

File: libs/types/src/tokens.rs (hunk 4 above)
Comment (lemunozm) [reply libs/types/src/tokens.rs:188]: Can we 👍🏻

Comment (lemunozm) [reply libs/types/src/tokens.rs:188]: Actually, after using the last cool bench utility `bench_default_investment_id()` (thanks @wischli for abstracting this), the default requirement in benchmarking is no longer needed. 

Thanks for pointing this out!

File: libs/test-utils/src/mocks/accountant.rs (hunk 5 above)
Comment (lemunozm) [reply libs/test-utils/src/mocks/accountant.rs:178]: Ideally, we could change this by the mock coming from `mock-builder`, but I had no time to refactor the tests with this

File: pallets/investments/src/benchmarking.rs [hunk 6]
```diff
@@ -10,3 +10,131 @@
...
+
+			Pallet::<T>::invest_fulfillment(investment_id, fulfillment)?;
+		}
+
+		#[extrinsic_call]
+		collect_investments(RawOrigin::Signed(caller), investment_id);
+	}
+
+	#[benchmark]
+	fn collect_redemptions(n: Linear<1, 10>) {
```
Comment (lemunozm) [inline pallets/investments/src/benchmarking.rs:112]:
```suggestion
	fn collect_redemptions(n: Linear<1, { T::MaxOutstandingCollects::get() }>) {
```

Debt-tech once we upgrade to Polkadot `v0.9.39`

**Review by mustermeiszer — APPROVED** [review]

**Pull request merged**